src/st/np3/44DCC.c

```diff
@@ -0,0 +1,1319 @@
+
+// EntitySurfacingWater ID 0x2E
+INCLUDE_ASM("asm/us/st/np3/nonmatchings/44DCC", func_801C56D8);
+
+// EntitySideWaterSplash ID 0x2F
+INCLUDE_ASM("asm/us/st/np3/nonmatchings/44DCC", func_801C5BC4);
+
+// EntitySmallWaterDrop ID 0x30
+INCLUDE_ASM("asm/us/st/np3/nonmatchings/44DCC", func_801C5F58);
+
+// EntityWaterDrop ID 0x31
+INCLUDE_ASM("asm/us/st/np3/nonmatchings/44DCC", func_801C61B4);
+
+s32 func_801C6458(s16 yVector) {
+    s16 newY = yVector + g_CurrentEntity->posY.i.hi;
+    s32 expectedResult = 0;
+    Collider collider;
+    Entity* newEntity;
+    s32 res;
+
+    g_api.CheckCollision(g_CurrentEntity->posX.i.hi, newY, &collider, 0);
+    res = expectedResult == (collider.unk0 & 1);
+
+    if (collider.unk0 & 8) {
+        if (*(u8*)&g_CurrentEntity->ext.generic.unkA0 == 0) {
+            newEntity = AllocEntity(&D_8007DE38, &D_8007DE38[24]);
+            if (newEntity != NULL) {
+                CreateEntityFromEntity(0x33, g_CurrentEntity, newEntity);
+                newEntity->posY.i.hi += yVector;
+                newEntity->zPriority = g_CurrentEntity->zPriority;
+            }
+            g_api.PlaySfx(0x7C2);
+            *(u8*)&g_CurrentEntity->ext.generic.unkA0 = 1;
+        }
+    }
+    return res;
+}
+
+// Matching, assembler skips a NOP
+// https://decomp.me/scratch/8wa56
+#ifndef NON_MATCHING
+INCLUDE_ASM("asm/us/st/np3/nonmatchings/44DCC", EntityMerman2);
+#else
+void EntityMerman2(Entity* self) {
+    s32 firstPrimIndex;
+    Collider collider;
+    Entity* newEntity;
+    Entity* newEntity2;
+    Primitive* prim;
+    s16 posX, posY;
+    s32 colRes;
+    s32 offset;
+    s32 index;
+    u32 accel;
+    s16* pos;
+    s16 camY;
+    s32 rnd;
+    s32 res;
+    s32 i;
+
+    if (self->ext.merman2.ignoreCol && (self->step < MERMAN2_7)) {
+        SetStep(MERMAN2_7);
+    }
+
+    if ((self->flags & 0x100) && (self->step < MERMAN2_DYING)) {
+        func_801C2598(0x71D);
+        self->unk19 = 0;
+        if (self->flags & FLAG_FREE_POLYGONS) {
+            g_api.FreePrimitives(self->firstPolygonIndex);
+            self->flags &= ~FLAG_FREE_POLYGONS;
+        }
+        self->unk3C = 0;
+        self->flags &= ~FLAG_UNK_20000000;
+        SetStep(MERMAN2_DYING);
+    }
+
+    switch (self->step) {
+    case MERMAN2_INIT:
+        InitializeEntity(D_80180AB4);
+        self->unk12 = 8;
+        self->zPriority = 0xA9;
+        if (self->subId != 0) {
+            self->palette++;
+        }
+        self->accelerationY = -0x10000;
+        break;
+
+    case MERMAN2_SWIMMING_UP:
+        AnimateEntity(D_8018229C, self);
+        MoveEntity();
+        posY = self->posY.i.hi - 24;
+        g_api.CheckCollision(self->posX.i.hi, posY, &collider, 0);
+        if (!(collider.unk0 & 8)) {
+            SetStep(MERMAN2_SWIMMING);
+            break;
+        }
+
+    case MERMAN2_SWIMMING:
+        if (self->step_s == 0) {
+            rnd = (Random() & 3) * 2;
+            self->accelerationX = D_80182224[rnd];
```

Comment:
It didn't match with unkstruct6 because it was composed of two s16s and the function needs s32s, so i made another struct and renamed the previous one. Point16 and Point32 respectively